commands/audit/sca/pnpm/pnpm.go

```diff
@@ -0,0 +1,168 @@
+
+import (
+	"encoding/json"
+	"errors"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/jfrog/gofrog/datastructures"
+	"github.com/jfrog/gofrog/io"
+	"github.com/jfrog/jfrog-cli-core/v2/utils/coreutils"
+	"golang.org/x/exp/maps"
+
+	"github.com/jfrog/jfrog-cli-security/utils"
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"github.com/jfrog/jfrog-client-go/utils/io/fileutils"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+
+	coreXray "github.com/jfrog/jfrog-cli-core/v2/utils/xray"
+	xrayUtils "github.com/jfrog/jfrog-client-go/xray/services/utils"
+)
+
+type pnpmLsDependency struct {
+	From         string                      `json:"from"`
+	Version      string                      `json:"version"`
+	Dependencies map[string]pnpmLsDependency `json:"dependencies,omitempty"`
+}
+
+type pnpmLsProject struct {
+	Name            string                      `json:"name"`
+	Version         string                      `json:"version"`
+	Dependencies    map[string]pnpmLsDependency `json:"dependencies,omitempty"`
+	DevDependencies map[string]pnpmLsDependency `json:"devDependencies,omitempty"`
+}
+
+func BuildDependencyTree(params utils.AuditParams) (dependencyTrees []*xrayUtils.GraphNode, uniqueDeps []string, err error) {
+	// Prepare
+	currentDir, err := coreutils.GetWorkingDirectory()
+	if err != nil {
+		return
+	}
+	pnpmExecPath, err := getPnpmExecPath()
+	if err != nil {
+		return
+	}
+	// Build
+	if err = installProjectIfNeeded(pnpmExecPath, currentDir); errorutils.CheckError(err) != nil {
+		return
+	}
+	return calculateDependencies(pnpmExecPath, currentDir, params)
+}
+
+func getPnpmExecPath() (pnpmExecPath string, err error) {
+	if pnpmExecPath, err = exec.LookPath("pnpm"); errorutils.CheckError(err) != nil {
+		return
+	}
+	if pnpmExecPath == "" {
+		err = errors.New("could not find the 'pnpm' executable in the system PATH")
+		return
+	}
+	log.Debug("Using Pnpm executable:", pnpmExecPath)
+	// Validate pnpm version command
+	version, err := getPnpmCmd(pnpmExecPath, "", "--version").RunWithOutput()
+	if errorutils.CheckError(err) != nil {
+		return
+	}
+	log.Debug("Pnpm version:", string(version))
+	return
+}
+
+func getPnpmCmd(pnpmExecPath, workingDir, cmd string, args ...string) *io.Command {
+	command := io.NewCommand(pnpmExecPath, cmd, args)
+	if workingDir != "" {
+		command.Dir = workingDir
+	}
+	return command
+}
+
+// Install is required when "pnpm-lock.yaml" lock file or "node_modules/.pnpm" directory not exists.
+func installProjectIfNeeded(pnpmExecPath, workingDir string) (err error) {
+	lockFileExists, err := fileutils.IsFileExists(filepath.Join(workingDir, "pnpm-lock.yaml"), false)
+	if err != nil {
+		return
+	}
+	pnpmDirExists, err := fileutils.IsDirExists(filepath.Join(workingDir, "node_modules", ".pnpm"), false)
+	if err != nil || (lockFileExists && pnpmDirExists) {
+		return
+	}
+	// Install is needed
+	log.Debug("Installing Pnpm project:", workingDir)
+	return getPnpmCmd(pnpmExecPath, workingDir, "install").GetCmd().Run()
```

Comment:
Please consider adding the [--ignore-scripts](https://pnpm.io/cli/install#--ignore-scripts) flag to block remote code execution attempts in Frogbot.
See the `addIgnoreScriptsFlag` function in `npm.go`.

Reply:
In Pnpm `ls` there is no `--ignore-scripts` flag
![image](https://github.com/jfrog/jfrog-cli-security/assets/49212512/71c0fff1-a537-472d-8d61-a647b1447810)